File: src/readctval.c
/*
 * NAME:
 * readctval.c
 *
 * PURPOSE:
 * To dump data stored in NCSA HDF5 files by software collocate...
 *
 * REQUIREMENTS:
 * NA
 *
 * INPUT:
 * NA
 *
 * OUTPUT:
 * NA
 *
 * NOTES:
 * NA
 *
 * BUGS:
 * NA
 *
 * AUTHOR:
 * Øystein Godøy, MET/FOU, 21.01.2003
 *
 * MODIFIED:
 * NA
 *
 * CVS_ID:
 * $Id: readctval.c,v 1.1 2007-09-07 17:03:00 steingod Exp $
 */

#ifdef HAVE_LIBHDF5

#include <readctval.h>
#include <unistd.h>
#include <dirent.h>
#include <time.h>
#include <utime.h>
#include <sys/types.h>
#include <sys/stat.h>

int sm_debug=1;

void readctval(char **infile, int *n, int *p, char **station, 
	int *start, int *end,
	int *stid, int *mytime,
	int *N, int *CL, int *CM, int *CH, 
	int *E, int *sss, int *cm) {

    /* General variables */
    char *where="readctval";
    int i, j, k;
    int status, noobs;
    int allmissing=0;
    obsstruct *sobs=NULL;
    as_data *asdata=NULL;
    ns_data *nsdata=NULL;
    safcm_data2 *cmdata=NULL;
    safssi_data *ssidata=NULL;
    /*skeys scrit = {"CTVAL","NA",0,0};*/
    skeys scrit = {"cloud","NA",0,0};
    struct tm ut;
    time_t vt, mymax=0, mymin=INT_MAX;

    if (*p != BOXSIZE2D) {
	error(where,"BOXSIZE differs...");
	return;
    }

    sprintf(scrit.station,"%s",*station);
    scrit.t_start = (time_t) *start;
    scrit.t_end = (time_t) *end;
    printf(">> %d %d\n",(int) scrit.t_start,(int) scrit.t_end);

    if ((noobs=read34(*infile, &scrit, &sobs, &cmdata)) <= 0) {
	error(where," File trouble...");
    }
    if (noobs != *n) {
	error(where,"Allocated space differs...");
	printf(" %d %d\n", noobs, *n);
	return;
    }
    printf("Data read from HDF5 file...\n");

    /*
     * Retrieve data for analysis within R
     */
    j=0;

    for (k=0;k<noobs;k++) {
	ut.tm_year = ((sobs[k].year)-1900);
	ut.tm_mon = ((sobs[k].month)-1);
	ut.tm_mday = sobs[k].day;
	ut.tm_hour = sobs[k].hour;
	ut.tm_min = sobs[k].min;
	ut.tm_sec = 0;
	ut.tm_isdst = 0;
	vt = mktime(&ut);
	/*
	printf("%03d %d%02d%02d %02d:%02d - %d\n",
		k,
		sobs[k].year,sobs[k].month,sobs[k].day,
		sobs[k].hour,sobs[k].min,
		vt);
	*/
	if (vt<mymin) {
	    mymin=vt;
	}
	if (vt>mymax) {
	    mymax=vt;
	}
	if (sobs[k].N < 0 || sobs[k].N > 8) {
	    (*n)--;
	    continue;
	}
	if (cmdata[k].nopix < 0 || cmdata[k].nopix > 169) {
	    (*n)--;
	    continue;
	}
	for (i=0;i<cmdata[k].nopix;i++) {
	    stid[j] = atoi(sobs[k].stID);
	    mytime[j] = (int) vt;
	    N[j] = (int) sobs[k].N;
	    CL[j] = (int) sobs[k].Cl;
	    CM[j] = (int) sobs[k].Cm;
	    CH[j] = (int) sobs[k].Ch;
	    E[j] = (int) sobs[k].E;
	    sss[j] = (int) sobs[k].sss;
	    cm[j] = (int) cmdata[k].data[i];
	    if (cmdata[k].data[i] == 0) allmissing++;
	    j++;
	}
	if (allmissing == cmdata[k].nopix) {
	    printf("No valid pixels in this tile\n");
	    printf(" %d", j);
	    j -= allmissing;
	    printf(" %d\n", j);
	    (*n)--;
	}
	allmissing = 0;
    }
    printf(" Max utime: %d\n Min utime: %d\n",(int) mymax,(int)mymin);

    if (cmdata != NULL) {
	printf("Freeing FMCOL's CMDATA data structure.\n");
	free(cmdata);
    }
    if (sobs != NULL) {
	printf("Freeing FMCOL's SYNOP data structure.\n");
	free(sobs);
    }

    printf(" noobs: %d n: %d\n", noobs, (*n));

    return;
}

#endif

File: R/readctval.R
#
# NAME:
# readctval
#
# PURPOSE:
# To read collocate HDF4 files containing validation data for SAFNWC CM.
#
# NOTES:
# Not fully finished... 
#
# BUGS:
# Probably in the underlying C-function...
# C functions does not transfer strings back properly...
#
# CVS_ID:
# $Id: readctval.R,v 1.2 2008-04-29 16:28:44 steingod Exp $
#
# AUTHOR:
# Øystein Godøy, MET/FOU, 30.01.2003
#

readctval <- function(filename,classname="cloud",station="NA",start="NA",end="NA") {

    if (missing(filename)) {
	cat("Husk at filnavn må oppgis...\n")
	return;
    }

    start <- as.POSIXct(strptime(start,format="%d%b%Y"))
    end <- as.POSIXct(strptime(end,format="%d%b%Y"))
    if (is.na(start)) start <- 0
    if (is.na(end)) end <- 0

    noobs <- 0

    tmp <- .C("checkrec",
	filename=as.character(filename),
	noobs=as.integer(noobs),
	classname=as.character(classname),
	station=as.character(station),
	start=as.integer(start),end=as.integer(end))
    #dyn.unload("/home/steingod/software/R-functions/dumpcol/checkrec.so")

    if (tmp$noobs <= 0) return
    cat(paste("Fant:",tmp$noobs,"lagrede enheter\n"))

    nopix <- 13*13
    size <- tmp$noobs*nopix
    stid <- vector(mode="integer",length=size)
    tid <- vector(mode="integer",length=size)
    N <- vector(mode="integer",length=size)
    CL <- vector(mode="integer",length=size)
    CM <- vector(mode="integer",length=size)
    CH <- vector(mode="integer",length=size)
    E <- vector(mode="integer",length=size)
    sss <- vector(mode="integer",length=size)
    cm <- vector(mode="integer",length=size)
    
    #dyn.load("/home/steingod/software/R-functions/dumpcol/readctval.so")
    tmp <- .C("readctval",
	filename=as.character(filename),
	noobs=as.integer(tmp$noobs),nopix=as.integer(nopix),
	station=as.character(station),
	start=as.integer(start),end=as.integer(end),
	stid=as.integer(stid),tid=as.integer(tid),
	N=as.integer(N),
	CL=as.integer(CL), CM=as.integer(CM), CH=as.integer(CH),
	E=as.integer(E), sss=as.integer(sss),
	cm=as.integer(cm))

    cat(paste("Number of records found:",tmp$noobs,"\n"))
    if (tmp$noobs <= 0) {
	return(cat("Bailing out...\n"))
    }
    cat(paste("Fant:",tmp$noobs,"lagrede enheter etter forkastning\n"))
    validata <- tmp$noobs*169

    tmp$tid <- ISOdate(1970,1,1)+tmp$tid 
    return(data.frame(
	stid=tmp$stid[1:validata],
	noobs=tmp$noobs[1:validata],
	tid=tmp$tid[1:validata],
	N=tmp$N[1:validata],
	CL=tmp$CL[1:validata],
	CM=tmp$CM[1:validata],
	CH=tmp$CH[1:validata],
	E=tmp$E[1:validata],
	sss=tmp$sss[1:validata],
	cm=tmp$cm[1:validata]
	))
}
